### DISCUSSION TITLE
[[OUTDATED]] Nominal value saving and restoring utility

### DISCUSSION THREAD

**Discussion by klauer** [pcdshub/atef#4] [Ideas]
Ref #3 

ATEF should include a way to restore nominal values of configuration PVs

**Comment by klauer** [comment]
Maybe:
   - Record device state by way of `read()/read_configuration()`
   - Restore by way of list of attribute names? `restore_attrs = ["sig1", "cpt.sig2"]`

**Reply by klauer** [reply]
@ZLLentz wrote here https://github.com/pcdshub/atef/discussions/3#discussioncomment-560350
> Perhaps we can limit this sort of interaction to very targeted restores e.g.
> 
> ```
> $ atef restore im1l0
> Please confirm the following changes:
> Field      Old New
> Event Code 45  140
> ```

**Reply by klauer** [reply]
(7 months later)

What if this restore interface were a TUI/GUI with the ability to restore what you select?

**Comment by klauer** [comment]
What is there to learn from existing save/restore utilities? What is the overlap of those utilities and ATEF's restore functionality?

**Reply by ZLLentz** [reply]
Some save/restore utilities I know about:
- Autosave
  - Brings values back after an IOC restart
  - Only restores the most recent values, and in a text file- easy to break if it fails once
  - Happens at IOC startup and can be interfered with by race conditions
- Parameter Manager
  - Useful but somewhat complicated
  - Configs are defined separately from devices and assigned to devices
  - Backed by an SQL database
- The thing ACR uses to repeatedly put to PVs
  - I don't know much about this

**Reply by klauer** [reply]
ACR tool -> SCORE perhaps?
<img width="472" alt="image" src="https://user-images.githubusercontent.com/5139267/139965389-72aebc86-f29e-469f-a59d-f7c9704b1413.png">

**Reply by ZLLentz** [reply]
Yes, that's what I was thinking about

**Reply by ZLLentz** [reply]
Compared to these existing tools, I think ATEF needs to have a quick user-friendly way to see history for (archiver + saved configuration) and restore to individual values
